### ISSUE TITLE
Marker which wraps more than one block element is not moved to the $graveyard

### ISSUE THREAD

**Issue by oskarwrobel** [ckeditor/ckeditor5-engine#968]
It doesn't work for the following scenario:
```javascript
// Set data and selection.
setData( document, '<paragraph>f[oo bar</paragraph><paragraph>foo ba]r</paragraph>' );

// Create marker <paragraph>foo b[ar</paragraph><paragraph>fo]o bar</paragraph>.
document.markers.set( 'something:1', new Range(
	new Position( root, [ 1, 5 ] ),
	new Position( root, [ 2, 2 ] )
) );

// Remove selected text (with marker inside).
editor.execute( 'delete' );

// Marker range is still in the main root.
console.log( document.marker.get( 'something:1' ).getRange().root.rootName ) // main
```

This one works fine:
```javascript
// Set data and selection.
setData( document, '<paragraph>f[oo ba]r</paragraph>' );

// Create marker <paragraph>fo[o b]ar</paragraph>.
document.markers.set( 'something:1', new Range(
	new Position( root, [ 1, 2 ] ),
	new Position( root, [ 1, 5 ] )
) );

// Remove selected text (with marker inside).
editor.execute( 'delete' );

// Marker range is in $graveyard.
document.markers.get( 'something:1' ).getRange().root.rootName; // $graveyard
```

**Comment by Reinmar** [comment]
And what implications does it have? I'm trying to understand the priority.

**Comment by oskarwrobel** [comment]
There is no information that such a marker is removed. Removed marker (together with text) is still present in the `MarkersCollection` (what is correct) and `root` of the marker range is still `main` (what is incorrect).

When this marker represents something that should be removed together with it (e.g. a comment), then information about that is missing.

**Comment by scofalik** [comment]
This is because how ranges works and how `DeleteCommand` works.

Assume that `{}` is marker and `[]` is removed range. This is what happens:

1. `<p>Lor[em i{psum</p><p>dol}or s]it amet</p>`.
2. `<p>Lor[em i{psum</p><p>}it amet</p>`.
3. `<p>Lor{</p><p>}it amet</p>`.
4. `<p>Lor{}it amet</p><p></p>`.
5. `<p>Lor{}it amet</p>`.

Unfortunately this will be hard to overcome by some generic algorithm so I am afraid it will be best tackle it by a special treating during `DeleteCommand` (we can hook to proper `DeleteCommand#event:execute`). Though I really, really don't like such solutions.

**Comment by Reinmar** [comment]
Couldn't we make this an option of a marker whether it gets moved to the clipboard once it gets collapsed? This can, of course, be also managed from outside (by the code which created markers)

I don't think that if you made a non-collapsed marker you'd like to keep it when it's collapsed.

**Comment by scofalik** [comment]
What exactly happens is that before merging, marker starts at the end of first element and ends between merged elements: `<p>foo{</p>}<p></p>`. Then, next element, is moved to the graveyard and the marker stays as it is. The marker is not even collapsed, and we cannot check whether it `isEmpty` because this method should work during OT (where we don't have tree model).

This is a very tricky case.

**Comment by scofalik** [comment]
I think that for now we should try to "hack" it using `DeleteCommand` execute event.

**Comment by Reinmar** [comment]
I still don't understand why the code which adds markers to the document cannot control that?

Can't it simply remove markers once their start starts touching their end?

**Comment by scofalik** [comment]
Yes, you can do that, though I think that if you want to do this like this, it's better to remove the marker in `deleteContent`.

If we will do it like you propose, it will be unable to undo the marker to the state before it was removed (because you have a collaps-y marker at some point and it won't uncollapse).

But all I wanted to note is that all transformations to any range should be done in any method (`Range#_getTransformedByDocumentChange`) so all algorithms that need to transform ranges by document changes will use it (undo, markers, `LiveRange`s, document selection).

**Comment by scofalik** [comment]
After all, I think that it's even nice that the code that "fixes markers behaviour" lands in `deleteContent`. It's nice because it's `deleteContent` what brings the problem at the first place (because it deletes stuff in three steps). So, I think that for now this will be a fine place to fix this.

**Comment by scofalik** [comment]
Unfortunately, it won't work. Even if the marker is removed through `MarkerOperation`, it won't work on undo. Because the (reversed) `MarkerOperation`, on undo, will get transformed through `RemoveOperation`s and will get shrunk anyway.

**Comment by Reinmar** [comment]
OK, it seems as a too big task for now, so I'm removing this from the current milestone.

**Comment by Reinmar** [comment]
Removed from 1.0.0 candidates because we think that it's not feasible to fix it.
